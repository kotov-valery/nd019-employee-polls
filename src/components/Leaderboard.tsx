import { UserList } from "../backend/Types";

function Leaderboard({ userList }: { userList: UserList | null }) {
  if (!userList) {
    return <div>Loading...</div>;
  }

  const sortedUsers = Object.values(userList).sort((a, b) => {
    const aScore = Object.keys(a.answers).length + a.questions.length;
    const bScore = Object.keys(b.answers).length + b.questions.length;
    return bScore - aScore;
  });

  return (
    <div className="leaderboard">
      <h1>Leaderboard</h1>
      <table className="leaderboard-table">
        <thead>
          <tr>
            <th>Rank</th>
            <th>User</th>
            <th>Answers</th>
            <th>Questions</th>
          </tr>
        </thead>
        <tbody>
          {sortedUsers.map((user, index) => (
            <tr key={user.id}>
              <td>{index + 1}</td>
              <td>{user.name}</td>
              <td>{Object.keys(user.answers).length}</td>
              <td>{user.questions.length}</td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}

export default Leaderboard;
