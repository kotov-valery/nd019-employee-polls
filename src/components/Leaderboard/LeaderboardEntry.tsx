import { User } from "../../backend/Types";

function LeaderboardEntry({ user, index }: { user: User; index: number }) {
  return (
    <tr key={user.id}>
      <td>{index + 1}</td>
      <td>
        <div className="leaderboard-user-entry">
          <img
            className="leaderboard-user-avatar"
            src={user.avatarURL}
            alt={`${user.name}'s avatar`}
          />
          <div className="leaderboard-user-text">
            <span className="leaderboard-user-id">{user.id}</span>
            <span className="leaderboard-user-name">{user.name}</span>
          </div>
        </div>
      </td>
      <td>{Object.keys(user.answers).length}</td>
      <td>{user.questions.length}</td>
    </tr>
  );
}
export default LeaderboardEntry;
