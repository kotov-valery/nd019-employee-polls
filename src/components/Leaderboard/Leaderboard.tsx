import { UserList } from "../../backend/Types";
import LeaderboardEntry from "./LeaderboardEntry";

function Leaderboard({ userList }: { userList: UserList | null }) {
  if (!userList) {
    return <div>Loading...</div>;
  }

  const sortedUsers = Object.values(userList).sort((a, b) => {
    const aScore = Object.keys(a.answers).length + a.questions.length;
    const bScore = Object.keys(b.answers).length + b.questions.length;
    return bScore - aScore;
  });

  return (
    <div className="leaderboard">
      <h1>Leaderboard</h1>
      <table className="leaderboard-table">
        <thead>
          <tr>
            <th>Rank</th>
            <th>User</th>
            <th>Answers</th>
            <th>Questions</th>
          </tr>
        </thead>
        <tbody>
          {sortedUsers.map((user, index) => (
            <LeaderboardEntry key={user.id} user={user} index={index} />
          ))}
        </tbody>
      </table>
    </div>
  );
}

export default Leaderboard;
