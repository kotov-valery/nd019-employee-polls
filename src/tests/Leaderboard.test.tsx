import { render } from "@testing-library/react";
import Leaderboard from "../components/Leaderboard/Leaderboard";

test("Leaderboard snapshot", () => {
  const users = {
    mtsamis: {
      id: "mtsamis",
      name: "Mike Tsamis",
      avatarURL: "/avatars/mtsamis.png",
      answers: { q1: "optionOne", q2: "optionTwo" },
      questions: ["q1", "q2"],
    },
    johndoe: {
      id: "johndoe",
      name: "John Doe",
      avatarURL: "/avatars/johndoe.png",
      answers: { q3: "optionOne" },
      questions: ["q3"],
    },
  };

  const { asFragment } = render(<Leaderboard userList={users} />);
  expect(asFragment()).toMatchSnapshot();
});
