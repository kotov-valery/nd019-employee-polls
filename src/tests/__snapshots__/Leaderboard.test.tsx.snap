// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leaderboard snapshot 1`] = `
<DocumentFragment>
  <div
    class="leaderboard"
  >
    <h1>
      Leaderboard
    </h1>
    <table
      class="leaderboard-table"
    >
      <thead>
        <tr>
          <th>
            Rank
          </th>
          <th>
            User
          </th>
          <th>
            Answers
          </th>
          <th>
            Questions
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            1
          </td>
          <td>
            <div
              class="leaderboard-user-entry"
            >
              <img
                alt="Mike Tsamis's avatar"
                class="leaderboard-user-avatar"
                src="/avatars/mtsamis.png"
              />
              <div
                class="leaderboard-user-text"
              >
                <span
                  class="leaderboard-user-id"
                >
                  mtsamis
                </span>
                <span
                  class="leaderboard-user-name"
                >
                  Mike Tsamis
                </span>
              </div>
            </div>
          </td>
          <td>
            2
          </td>
          <td>
            2
          </td>
        </tr>
        <tr>
          <td>
            2
          </td>
          <td>
            <div
              class="leaderboard-user-entry"
            >
              <img
                alt="John Doe's avatar"
                class="leaderboard-user-avatar"
                src="/avatars/johndoe.png"
              />
              <div
                class="leaderboard-user-text"
              >
                <span
                  class="leaderboard-user-id"
                >
                  johndoe
                </span>
                <span
                  class="leaderboard-user-name"
                >
                  John Doe
                </span>
              </div>
            </div>
          </td>
          <td>
            1
          </td>
          <td>
            1
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
